src/lib/es2018.asyncgenerator.d.ts

```diff
@@ -2,8 +2,8 @@
 
 interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, TReturn, TNext> {
     // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
-    next(...args: [] | [TNext | PromiseLike<TNext>]): Promise<IteratorResult<T, TReturn>>;
-    return(value: TReturn | PromiseLike<TReturn>): Promise<IteratorResult<T, TReturn>>;

```

Comment:
Async generators actually do unwrap the value passed to `return`:

Example:
```ts
async function *gen() {
  yield 1;
  yield 2;
  return 3;
}
const a = gen();
a.return(Promise.resolve(-1)).then((result) => {
  // promise passed to return is unwrapped.
  console.log(result); // {value: -1, done: true}
});
```

This is described in [section 25.5.3.5.1](https://tc39.es/ecma262/#async-generator-resume-next-return-processor-fulfilled) of the tc39 spec.

So the former definition is more correct. However, I would argue that the above typing should not be applied to async iterators. The spec does not fully specify how the [AsyncIterator interface](https://tc39.es/ecma262/#sec-asynciterator-interface) should handle arguments to `return`. It states:

> The returned promise will fulfill with an IteratorResult object which will typically have a done property whose value is true, and a value property with the value passed as the argument of the return method.

Additionally the spec suggests that the value of the returned iterator result should not be a promise. Together, these two requirements imply that the return value is unwrapped if it’s a promise, but neither of these statements are normative; none of these requirements are enforced in any way.

Reply:
Unfortunately we can't have it both ways. `AsyncIterator` needs to be conformant to not only what you might produce with a custom iterator, but also the async iterators that are consumed by users from native APIs (async generators, NodeJS streams, DOM apis, etc.), so we need to conform to the strictest API between the two.

Reply:
Do you have an example of a DOM api which has extended AsyncIterator? I’ve looked in https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.iterable.d.ts, https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2018.asyncgenerator.d.ts, https://github.com/microsoft/TypeScript/blob/master/lib/lib.es2018.asynciterable.d.ts, https://github.com/microsoft/TypeScript/blob/master/lib/lib.esnext.asynciterable.d.ts but none of these files define the async iterator stuff for DOM types yet.

Reply:
So what you’re saying is that we can’t have a stricter interface for `AsyncGenerator` than for `AsyncIterator` like so:
```ts
interface AsyncGenerator<T = unknown, TReturn = any, TNext = unknown> extends AsyncIterator<T, PromiseLike<TReturn> | TReturn, TNext> {
    // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
    next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
    return(value: TReturn | PromiseLike<TReturn>): Promise<IteratorResult<T, TReturn>>;
    throw(e: any): Promise<IteratorResult<T, TReturn>>;
    [Symbol.asyncIterator](): AsyncGenerator<T, TReturn, TNext>;
}

interface AsyncIterator<T, TReturn = any, TNext = undefined> {
    // NOTE: 'next' is defined using a tuple to ensure we report the correct assignability errors in all places.
    next(...args: [] | [TNext]): Promise<IteratorResult<T, TReturn>>;
    return?(value?: TReturn): Promise<IteratorResult<T, TReturn>>;
    throw?(e?: any): Promise<IteratorResult<T, TReturn>>;
}
```

and have no such constraint for AsyncIterator. I’m not 100% on the reason why, but I’m fine with leaving the types of `return` unchanged, and I trust you know better when you say we can’t have it both ways!

Reply:
There are none yet that I know of, but I know there are discussions going on about future DOM apis that may use async iteration.